feat(manifest): Establish core manifest system foundation

undead2146 opened the pull request:
This PR introduces the foundational manifest and publisher-driven content ecosystem.

The following changes are made:

1. **Core Models & Interfaces**: These define the data contract for the manifest system, including `GameManifest`, `ManifestFile`, and supporting enums `ManifestFileSourceType.cs`, `ContentType.cs` and `WorkspaceStrategy.cs` Along with the service contracts: `IManifestCache`, `IManifestProvider`, `IContentManifestBuilder`, and `IManifestGenerationService`.
2. **Service Implementations**: These services provide the concrete implementations for all manifest services:
    * `ManifestDiscoveryService`: Scans for and loads manifest files from embedded resources and the file system.
    * `ManifestCache`: A singleton, in-memory cache that acts as the central repository for all discovered manifests.
    * `ManifestProvider`: A facade over the cache, providing other services with access to manifest data.
    * `ContentManifestBuilder`: A builder for creating `GameManifest` objects.
    * `ManifestGenerationService`: Orchestrates the creation of new `GameManifest` instances for base games, mods, addons, patches, and standalone versions by leveraging `IContentManifestBuilder` to scan directories, configure metadata and dependencies, add files, and serialize the resulting manifest to JSON.
    * `ManifestInitializationService`: An `IHostedService` that ensures the manifest system is initialized on application startup.
3. **Infrastructure**: Integrates services into dependency injection via `ManifestModule`. It also adds custom exception types like `ManifestNotFoundException` and `ManifestValidationException` for error handling.
4. **Unit Tests**: Introduces unit tests for the `ManifestCache`, `ManifestDiscoveryService`, and `ContentManifestBuilder`.

##### **Commit 1: `feat(core): Add data models for manifest system`**

This commit establishes the foundational data contracts for the entire manifest system. It defines the schema for what a "manifest" is and all the supporting data structures required to describe a piece of content, its files, and its metadata.

* **`GenHub.Core/Models/Manifest/GameManifest.cs`**: This is the central data model, representing a single distributable content package. The `Id` property (e.g., "Steam.ZeroHour") serves as a unique key. The `Files` property, a `List<ManifestFile>`, contains every file required for the content to run. The `RequiredDirectories` list ensures the launcher creates necessary folder structures. The `Publisher` property holds a `PublisherInfo` object for metadata, while `InstallationInstructions` provides a guide for complex setups.

* **`GenHub.Core/Models/Manifest/ManifestFile.cs`**: This model represents a single file within the `GameManifest`. The `RelativePath` property defines its location within the game's root directory. `Size` and `Hash` (SHA256) are critical for the upcoming Validation System to verify file integrity. The most important property is `SourceType`, a `ManifestFileSourceType` enum, which dictates how the launcher should acquire this file when building a workspace. For remote files, the `DownloadUrl` property specifies where to fetch the content from.

* **`GenHub.Core/Models/Enums/ContentType.cs`**: This enum categorizes the type of content a `GameManifest` represents. For example, `BaseGame` is used for a manifest describing the original, unmodified game. `Mod` is for a total conversion, and `Patch` is for a set of balance changes. This allows the launcher to understand the nature of the content and apply different logic, such as determining dependencies.

* **`GenHub.Core/Models/Enums/ManifestFileSourceType.cs`**: This is the most critical enum for the workspace creation logic. It tells the launcher how to handle each `ManifestFile`. `LinkFromBase` instructs the launcher to create a symbolic link to the file in the user's base game installation, saving significant disk space. `CopyUnique` is for files specific to the mod that must be copied. `Download` is for optional or large files hosted remotely. `Generate` is for files that need to be created on the fly, like a patched executable or a configuration file.

* **Other Models (`PublisherInfo.cs`, `ContentDependency.cs`, etc.)**: These are simple data-carrying models that provide structured metadata within the `GameManifest`. `PublisherInfo` contains fields like `Name` and `Website`. `ContentDependency` allows a manifest to declare that it requires another manifest to be present (e.g., a sub-mod requiring a main mod).

##### **Commit 2: `feat(core): Add service interfaces for manifest system`**

This commit defines the abstract contracts for the services that will operate on the data models defined in the previous commit.

* **`GenHub.Core/Interfaces/Manifest/IManifestProvider.cs`**: This interface acts as the primary facade for the rest of the application to interact with the manifest system. Its main responsibility is to retrieve `GameManifest` objects. It defines `GetManifestAsync` methods that can take a `GameVersion` or `GameInstallation` object, abstracting away the logic of how and where the manifest is found (cache, embedded resources, etc.).

* **`GenHub.Core/Interfaces/Manifest/IManifestCache.cs`**: This interface defines the contract for a singleton, in-memory cache for `GameManifest` objects. It exposes methods like `AddOrUpdateManifest(GameManifest manifest)` to populate the cache and `GetManifest(string manifestId)`.

* **`GenHub.Core/Interfaces/Manifest/IContentManifestBuilder.cs`**: This interface defines a builder pattern for constructing `GameManifest` objects. It provides a chainable API with methods like `WithBasicInfo(...)`, `WithFile(...)`, and `WithPublisher(...)`. This is essential for the `ManifestGenerationService`, which will use this builder to create new manifests by scanning existing game directories.

* **`GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs`**: This interface defines the contract for a high-level service responsible for creating new `GameManifest` files. It will orchestrate the process of scanning a game directory, using the `IContentManifestBuilder` to assemble a `GameManifest` object, and then serializing it to a JSON file.

##### **Commit 3: `feat(manifest): Implement discovery, cache, provider, and generation services`**

This commit provides the concrete implementations for the interfaces defined in the previous commit. This is where the core logic of the manifest system resides.

* **`GenHub/Features/Manifest/ManifestProvider.cs`**: This class implements `IManifestProvider`. It first attempts to retrieve a manifest from the `IManifestCache`. If not found, it falls back to loading it from embedded application resources. It also contains security logic to validate manifest file paths.

* **`GenHub/Features/Manifest/ManifestCache.cs`**: This class implements `IManifestCache` using a `ConcurrentDictionary<string, GameManifest>` as its backing store. The use of `ConcurrentDictionary` ensures that the cache is thread-safe.

* **`GenHub/Features/Manifest/ManifestDiscoveryService.cs`**: This service is responsible for finding and loading all manifests at application startup. It scans for `.json` files in predefined directories (like a `Manifests` folder) and also discovers manifests embedded within the application's assemblies. It then uses the `IManifestCache` to store them.

* **`GenHub/Features/Manifest/ManifestInitializationService.cs`**: This is an `IHostedService` that orchestrates the startup process. When the application starts, the Host calls its `StartAsync` method, which in turn calls the `ManifestDiscoveryService` to populate the cache.

##### **Commit 4: `refactor(infra): Add manifest exceptions and DI module`**

This commit handles infrastructure concerns, including custom error handling and dependency injection setup.

* **`GenHub/Infrastructure/Exceptions/ManifestExceptions.cs`**: This file defines custom exception types like `ManifestNotFoundException` and `ManifestValidationException`.

* **`GenHub/Infrastructure/DependencyInjection/ManifestModule.cs`**: This is a dependency injection module that registers all the manifest-related services with the DI container. It registers the `ManifestCache` as a singleton, the `ManifestProvider` as a singleton, and the `ManifestInitializationService` as a hosted service.

##### **Commit 5: `test(manifest): Add unit tests for manifest system foundation`**

This commit adds unit tests for all the new services and logic.

* **`GenHub.Tests/Features/Manifest/*Tests.cs`**: These files contain Xunit tests for each of the new services. For example, `ManifestProviderTests.cs` uses Moq to mock the `IManifestCache` and verifies that the provider correctly retrieves manifests from the cache. `ManifestCacheTests.cs` tests the thread-safety and correctness of the cache's add and retrieve operations. T

This work is a prerequisite for all subsequent features, including validation, workspace management, and content installation.

ViTeXFTW commented:
I will start by saying this PR description is really nice to have to get a high level view of the PR before diving into the code change/additions

ViTeXFTW commented on GenHub/GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <param name="modName">Mod display name.</param>
+    /// <param name="modVersion">Mod version.</param>
+    /// <param name="targetGame">Target game type.</param>
+    /// <param name="baseGameDependencies">Required base game versions.</param>
```
Is this a list of game version(s) that the mod can work with?

undead2146 replied:
Upon further reflection, I think it is better if the different methods `CreateManifestAsync` for mod, patch and addon are consolidated under a single `CreateContentManifestAsync` methods. Since we `ContentType` will allow for 

```csharp
// In IManifestGenerationService.cs
    Task<IContentManifestBuilder> CreateContentManifestAsync(
        string contentDirectory,
        string contentId,
        string contentName,
        string contentVersion,
        ContentType contentType,
        GameType targetGame,
        params ContentDependency[] dependencies);

// In ManifestGenerationService.cs
    public async Task<IContentManifestBuilder> CreateContentManifestAsync(
        string contentDirectory,
        string contentId,
        string contentName,
        string contentVersion,
        ContentType contentType,
        GameType targetGame,
        params ContentDependency[] dependencies)
    {
        _logger.LogInformation("Creating content manifest for {ContentName} v{ContentVersion}", contentName, contentVersion);

        var builder = CreateBuilder()
            .WithBasicInfo(contentId, contentName, contentVersion)
            .WithContentType(contentType, targetGame)
            .WithMetadata($"Content manifest for {contentName}");

        foreach (var dep in dependencies)
        {
            builder.AddDependency(
                dep.Id,
                dep.Name,
                dep.DependencyType,
                dep.MinVersion,
                dep.MaxVersion,
                dep.CompatibleVersions,
                dep.IsRequired,
                dep.IsExclusive,
                dep.ConflictsWith,
                dep.InstallBehavior);
        }

        await builder.AddFilesFromDirectoryAsync(contentDirectory, ManifestFileSourceType.CopyUnique);
        return builder;
    }

```

ViTeXFTW commented on GenHub/GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <summary>
+    /// Creates a manifest builder for a mod.
+    /// </summary>
+    /// <param name="modDirectory">Path to the mod directory.</param>
```
Shouldn't we have a standard place where all mods, addons, patches, etc are stored and then give the user the option to change this base directory in the settings tab.

xezon commented on GenHub/GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs:
```diff
@@ -0,0 +1,77 @@
+    /// Creates a manifest builder for a mod.
+    /// </summary>
+    /// <param name="modDirectory">Path to the mod directory.</param>
+    /// <param name="modId">Unique mod identifier.</param>
```
I suggest use GUID for id's. There are universially understood and generating them is practically guaranteed unique.

https://en.wikipedia.org/wiki/Universally_unique_identifier

undead2146 replied:
I have consolidated patch, mod, and add-on manifest generation into a single method, 
`CreateContentManifestAsync()`. The method signature now accepts:

- `contentId`  
- `contentName`  
- `contentVersion`  
- `contentType`
- `targetGame`

These parameters are passed to `CreateBuilder().WithBasicInfo(contentId, contentName, contentVersion)`, 
which sets the resulting `GameManifest`’s ID to `contentId`.

The `contentId` will serve as the unique identifier for each piece of content. This PR does not 
include any concrete strategy for generating these IDs. In a future PR, we should decide whether 
to use GUIDs or a composite key (for example, `contentName`, `contentVersion`, and `contentType`) 
as our unique identifier scheme.

xezon reviewed (commented):
I like the clean presentation of the various definition types.

xezon commented on GenHub/GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <param name="targetContent">What this patch applies to.</param>
+    /// <param name="targetVersion">Version of the target content.</param>
+    /// <returns>A task that returns a configured manifest builder.</returns>
+    Task<IContentManifestBuilder> CreatePatchManifestAsync(string patchDirectory, string patchId, string patchName, string patchVersion, string targetContent, string targetVersion);
```
Maybe Patch can build on top of Mod? But we can just say Mod can build on top of Mod.

xezon replied:
I do not necessarily agree with this distinction. I am happy to hear more opinions on this.

A Patch is technically the same as a Mod, but it has conceptually different Goals. I agree that historically community patches had been rather barebones with just INI files in them, but that is because they had small Goals and small Ambitions. GeneralsGamePatch has larger Goals, and touches all parts of the original game, and technically it is like a Mod, but conceptually a Patch.

ViTeXFTW commented:
Looks good, will wait with approval until comments have been answered

undead2146 commented on GenHub/GenHub.Core/Models/Enums/ContentType.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// Map pack or additional content.
+    /// </summary>
+    MapPack,
```
In order to support content bundles, I created the `ContentBundle` and `BundleItem` models, along with `CreateContentBundleAsync` method.

``` csharp
public class ContentBundle
{
    public string Id { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Version { get; set; } = string.Empty;
    public PublisherInfo Publisher { get; set; } = new();
    public List<BundleItem> Items { get; set; } = new();
    public ContentMetadata Metadata { get; set; } = new();
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

public class BundleItem
{
    public string ContentId { get; set; } = string.Empty;
    public string PublisherId { get; set; } = string.Empty;
    public string RequiredVersion { get; set; } = string.Empty;
    public bool IsOptional { get; set; } = false;
    public ContentType ContentType { get; set; }
    public int DisplayOrder { get; set; } = 0;
}

// New method in IManifestGenerationService
// Bundle management
public async Task<ContentBundle> CreateContentBundleAsync(
    string bundleId,
    string bundleName,
    string bundleVersion,
    PublisherInfo publisher,
    params BundleItem[] items)
{
    _logger.LogInformation("Creating content bundle {BundleId} with {ItemCount} items", bundleId, items.Length);
    var bundle = new ContentBundle
    {
        Id = bundleId,
        Name = bundleName,
        Version = bundleVersion,
        Publisher = publisher,
        Items = items.OrderBy(i => i.DisplayOrder).ToList(),
        Metadata = new ContentMetadata
        {
            Description = $"Content bundle containing {items.Length} items",
            ReleaseDate = DateTime.UtcNow
        }
    };
    await ValidateBundleItemsAsync(bundle);
    return bundle;
}

```

**Concrete Example**

``` csharp
// Creating a comprehensive bundle
var megaPack = await manifestGeneration.CreateContentBundleAsync(
    "CommunityOutpost.MegaPack.v2.0",
    "Ultimate Community Mega Pack",
    "2.0",
    new PublisherInfo 
    { 
        Name = "Community Outpost",
        Website = "https://communityoutpost.com",
        ContactEmail = "support@communityoutpost.com"
    },
    new BundleItem
    {
        ContentId = "EnhancedMaps.Desert.v1.5",
        PublisherId = "GenTool",
        ContentType = ContentType.MapPack,
        RequiredVersion = "1.5",
        DisplayOrder = 1
    },
    new BundleItem
    {
        ContentId = "BalancePatch.Competitive.v1.3",
        PublisherId = "CommunityOutpost",
        ContentType = ContentType.Patch,
        RequiredVersion = "1.3",
        IsOptional = true,
        DisplayOrder = 2
    }
);

```

undead2146 commented:
My most recent commit contains the following changes:

**Unified Content Creation**  
- Removed `CreateModManifestAsync`, `CreatePatchManifestAsync`, and `CreateAddonManifestAsync` from [ManifestGenerationService.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-d7cc6713fe5041760db7040c9d22fa1471b009bb0c21c56e7184b3652afccdc8)  
- Introduced a single, unified method: [CreateContentManifestAsync](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-d7cc6713fe5041760db7040c9d22fa1471b009bb0c21c56e7184b3652afccdc8)

**Redesigned Dependency Model**  
- Expanded the [ContentDependency](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-69779175424cbbcf3e98d76949767bf84dfa65ba76abfd17737dc9fd5e655731) class with new members:  
  - `List<string> CompatibleVersions`  
  - `bool IsExclusive`  
  - `List<string> ConflictsWith`  
  - `DependencyInstallBehavior` enum  
- Updated [ContentManifestBuilder.AddDependency](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-d7cc6713fe5041760db7040c9d22fa1471b009bb0c21c56e7184b3652afccdc8) to accept these new parameters.

**Content Bundling & Referencing**  
- Added new models: [ContentBundle.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-6f9d1cb69571ec790c871cafdc02a1df45719532c9364624923f14f1e559d6c1) and [BundleItem.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-e8f087c1345161e221bbb2d70428041a31f38466ad4e93c85301c64b9744d611)  
- Added `CreateContentBundleAsync` method to [ManifestGenerationService.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-d7cc6713fe5041760db7040c9d22fa1471b009bb0c21c56e7184b3652afccdc8)  
- Extended [GameManifest.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-400aab7638fe203cd85022eb109731fc3fd47e483a7af6d56034836a367cab6e) with a `List<ContentReference> ContentReferences` property.  
- Added `AddContentReference` method to [ContentManifestBuilder.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-4deac36389576243d6f633054c420e09172abb4e52c4fa3baf6029cc7d186b38)

**Expanded Data Models**  
- Extended the [ContentType](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-c66c582c41d1f2a5165369f25342223e88305b57b4693b7312ca842e7459abdb) enum with `LanguagePack`, `ContentBundle`, `PublisherReferral`, and `ContentReferral`.  
- Added the `DependencyInstallBehavior` property to [ContentDependency.cs](https://github.com/community-outpost/GenHub/pull/60/commits/ce15ff45821b4b1d36030e9764f2103b6e39e608#diff-69779175424cbbcf3e98d76949767bf84dfa65ba76abfd17737dc9fd5e655731).
